File: matlab/EuglenaData.m
classdef EuglenaData < handle 
   properties (Access = public)
        tracks
        lightData
        metaData
        zoom
        timeTable
        lightTable
        trackID2IndexMap
        totalTime
        fps
        numFrames
   end
   
   methods (Access = public)
       function this  = EuglenaData(path)
            % EuglenaTracks(path,
            if path(end) ~= filesep || path(end) ~= '/'
                path = strcat(path,filesep);
            end
            
            tracksPath = strcat(path,'tracks.json');
            fid = fopen(tracksPath);
            raw = fread(fid,inf);
            str = char(raw');
            fclose(fid);
            this.tracks = fromjson(str);
                        
            lightDataPath = strcat(path,'lightdata.json');            
            fid = fopen(lightDataPath);
            raw = fread(fid,inf);
            str = char(raw');
            fclose(fid);
            data = fromjson(str);
            
            this.lightData = data.eventsToRun;
            this.metaData  = data.metaData;
            
            if isfield(this.metaData,'numFrames')
               this.numFrames = this.metaData.numFrames;
            else
               error('The lightdata.json is not updated, use newer version');
            end
                        
            this.zoom = int32(this.metaData.magnification);           
            this.totalTime = single(this.lightData{end}.time - this.lightData{1}.time) / 1000.0;
            this.fps = single(this.numFrames) / this.totalTime;
            
            this.timeTable = int32(zeros(length(this.tracks), 4 ));
            
            for i = 1:length(this.tracks)   
                this.timeTable(i,:) = [this.tracks{i}.trackID, this.tracks{i}.startFrame+1, this.tracks{i}.lastFrame+1, this.tracks{i}.numSamples];
            end            
            this.trackID2IndexMap = containers.Map( this.timeTable(:,1), 1:length(this.tracks) );
            
            this.lightTable = single(zeros(length(this.lightData),5));
                
            baseTime = 0;
            if ~isempty(this.lightData)
                baseTime = double(this.lightData{1}.time);
            end

            for i = 1:length(this.lightData)
                l = this.lightData{i};
                this.lightTable(i,:) = [ single((double(l.time) - (baseTime))/1000.0), single(l.topValue), single(l.rightValue), ...
                                         single(l.bottomValue), single(l.leftValue) ];
            end
       end
       
       function umpp = getUMPP(this)
           umpp = 100.0/30.0 * 4.0 / double(this.zoom);
       end
       
       function n = getNumTracks(this)
           n = length(this.tracks);
       end
       
       function time = getTotalTime(this)
           time = this.totalTime;
       end
       
       function mag = getMagnification(this)
           mag = this.zoom;
       end
       
       function fps = getFPS(this)
           fps = this.fps;
       end
       
       function idx = findTrackID(this,trackID)
           idx = this.trackID2IndexMap(trackID );
       end
       
       function filteredTracks = findTracksBetweenFrames(this,startFrame, endFrame)
            filter = ~(this.timeTable(:,3) < startFrame | this.timeTable(:,2) > endFrame);
            filteredTracks = this.tracks(filter);            
       end
       
       function track = getTrackAt(this,index)
           track = this.tracks{index};
       end
       
       function track = getTrackByID(this,id)
           idx = this.findTrackID(id);
           track = this.getTrackAt(idx);
       end
       
       function [x,y,width,height,angles,frames] = extractEuglenaFromFrames( this, startFrame, endFrame )
           
           x      = [];
           y      = [];
           width  = [];
           height = [];
           angles = [];
           frames = [];
           
           for i=1:this.getNumFrames()
               t = this.tracks{i};
               if ~(t.startFrame > endFrame || t.lastFrame < startFrame)
                   for j = 1:t.numSamples
                       s = t.samples{j};
                       if (s.frame) >= startFrame && (s.frame <= endFrame)
                           x(end+1)      = s.rect{1};
                           y(end+1)      = s.rect{2};
                           width(end+1)  = s.rect{3};
                           height(end+1) = s.rect{4};
                           angles(end+1) = s.rect{5};
                           frames(end+1) = s.frame;
                       end
                   end
               end
           end
           
       end
       
       function leds = getLedStatesFromTime( this, t )
           mask = t < this.lightTable(:,1);
           idx = find(mask);
           if isempty(idx)
               leds = [0,0,0,0];
               return
           end
           
           idx = idx(1);
           if idx == 1
               leds = [0,0,0,0];
               return
           else
               leds = this.lightTable(idx-1,2:end);
               return
           end
       end 
      
       function leds = getLedStatesFromFrame( this, frame )
           leds =  this.getLedStatesFromTime( (single(frame)-1) * 1.0/this.getFPS() );
       end
   end
   
   methods(Static)
       function allClippedTracks = clipTracksBetweenFrames(tracks, startFrame, endFrame)
           
           allClippedTracks = {};
           
           if ~iscell(tracks)
               tracks = {tracks};
           end
           
           for i = 1:length(tracks)
               t = tracks{i};
               if t.startFrame > endFrame || t.lastFrame < startFrame 
                   %completely outside, so just ignore                   
               elseif t.startFrame >= startFrame && t.lastFrame <= endFrame
                   %completely inside, so just take the whole thing
                   allClippedTracks{end+1} = t;
               else
                   clippedTrack = t;
                   clippedTrack.samples = {};
                   minFrame = inf;
                   maxFrame = -1;
                   for j = 1:t.numSamples
                       sample = t.samples{j};
                       if sample.frame >= startFrame && sample.frame <= endFrame

                           minFrame = min(minFrame,sample.frame);
                           maxFrame = max(maxFrame,sample.frame);                      

                           clippedTrack.samples{end+1} = sample;
                       end
                   end

                   if ~isempty(clippedTrack.samples) 
                       clippedTrack.startFrame = minFrame;
                       clippedTrack.lastFrame = maxFrame;
                       clippedTrack.numSamples = length(clippedTrack.samples);
                       allClippedTracks{end+1} = clippedTrack;
                   end
               end
           end
       end
       
       function [center_x,center_y,width,height,angles,frames] = extractTrackData(track)
           N = track.numSamples;
           
           center_x = zeros(1,N);
           center_y = zeros(1,N);
           angles   = zeros(1,N);
           frames   = zeros(1,N);
           width    = zeros(1,N);
           height   = zeros(1,N);           
           
           for i=1:N
               s = track.samples{i};
               rect = s.rect;
               center_x(i) = rect{1};
               center_y(i) = rect{2};
               width(i)    = rect{3};
               height(i)   = rect{4};
               angles(i)   = rect{5};
               frames(i)   = s.frame;
           end
       end
       
   end
end

File: python/EuglenaData.py
from __future__ import print_function
import sys

import os
import json
import copy
import csv
import numpy as np

def warning(*objs):
    print("WARNING: ", *objs, file=sys.stderr)

class EuglenaData(object):
    def __init__(self,path):
        if not path.endswith(os.pathsep):
            path += os.path.sep

        with open(path+"lightdata.json") as fp:
            self._lightData = json.load(fp)

        with open(path+"tracks.json") as fp:
            self._tracks = json.load(fp)

        ## Converting all frame numbers from 1 to N
        for i in xrange(len(self._tracks)):
            self._tracks[i]['startFrame'] += 1
            self._tracks[i]['lastFrame'] += 1
            samples = self._tracks[i]['samples']
            for j in xrange(len(samples)):
                samples[j]['frame'] += 1

        self._zoom = self._lightData['metaData']['magnification']
        self._events = self._lightData['eventsToRun']
        baseTime = self._events[0]['time']
        for i in xrange(len(self._events)):
            self._events[i]['time'] = (self._events[i]['time'] - baseTime) / 1000.0

        self._totalTime = (self._events[-1]['time'])
        self._trackIDtoIndexMap = { t['trackID']:i for i,t in enumerate(self._tracks) }

    def getUMPP(self):
        return 100.0/30.0 * 4.0 / float(self._zoom)


    def getNumTracks(self):
        return len(self._tracks)

    def getNumFrames(self):
        if "numFrames" in self._lightData['metaData']:
            return self._lightData['metaData']['numFrames']
        else:
            warning("Use newer version of data, numFrames information doesn't exists")
            return 0;

    def getTotalTime(self):
        return self._totalTime

    def getFPS(self):
        return self.getNumFrames() / self.getTotalTime()

    def getMagnification(self):
        return self._zoom

    def findTrackID(self,id):
        return self._trackIDtoIndexMap.get(id,-1)

    def findTracksBetweenFrames(self,startFrame, endFrame):
        return [ t for t in self._tracks if not (( t['startFrame']) > endFrame or (t['lastFrame']) < startFrame) ]


    def getTrackAt(self,index):
        return self._tracks[index]

    def getTrackByID(self,id):
        return self.getTrackAt( self.findTrackID(id) )

    def extractEuglenaBetweenFrames(self,startFrame,endFrame):
        x = []
        y = []
        w = []
        h = []
        a = []
        f = []

        for t in self._tracks:
            for s in t['samples']:
                if startFrame <=  (s['frame']) <= endFrame:
                    rect = s['rect']
                    x.append(rect[0])
                    y.append(rect[1])
                    w.append(rect[2])
                    h.append(rect[3])
                    a.append(rect[4])
                    f.append(s['frame'])

        return x,y,w,h,a,f

    def getLedStateFromTime(self,t):
        i = 0
        while i< len(self._events) and self._events[i]['time'] <= t:
            i += 1

        i -= 1
        if i >= 0:
            return self._events[i]['topValue'], self._events[i]['rightValue'], self._events[i]['bottomValue'], self._events[i]['leftValue']

        return 0,0,0,0

    def getLedStateFromFrame(self,frame):
        return self.getLedStateFromTime( frame / self.getFPS() )

    @staticmethod
    def clipTracksBetweenFrames(tracks,startFrame, endFrame):
        clippedTracks = []
        for t in tracks:
            if not (t['startFrame']) > endFrame or (t['lastFrame']) < startFrame:
                if (t['startFrame']) >= startFrame  and (t['lastFrame']) <= endFrame:
                    #completely inside
                    clippedTracks.append( copy.deepcopy(t) )
                else:
                    newTrack = copy.deepcopy(t)
                    newTrack['samples'] =filter(lambda ss: startFrame <= (ss['frame']) <= endFrame, t['samples'] )
                    clippedTracks.append(newTrack)
            else:
                newTrack = copy.deepcopy(t)
                newTrack['samples'] = []
                clippedTracks.append(newTrack)

        return clippedTracks

    @staticmethod
    def extractTrackData(track):
        # returns x,y,w,h,a,f
        return zip(*[(s['rect'][0], s['rect'][1], s['rect'][2], s['rect'][3], s['rect'][4], s['frame']) for s in track['samples'] ])

    def writeTrackDataToCSV(self,track,csvfile):
        x,y,w,h,a,f = np.array(EuglenaData.extractTrackData(track))

        x *= self.getUMPP()
        y *= self.getUMPP()
        w *= self.getUMPP()
        h *= self.getUMPP()

        t = f / self.getFPS()

        ledStates = [self.getLedStateFromFrame(fr) for fr in f]

        with open(csvfile,'wb') as fb:
            header = ('frame#','time (sec)','center_x (um)','center_y (um)','width (um)','height (um)','angle (degrees)','topLED','rightLED','bottomLED','leftLED')
            wr = csv.writer(fb)
            rows = zip(f,t,x,y,w,h,a)
            wr.writerow(header)
            for i,r in enumerate(rows):
                data = list(r)
                data.extend(ledStates[i])
                print(data)
                wr.writerow(data)
